src/hotspot/share/opto/parse1.cpp

```diff
@@ -1782,13 +1779,33 @@ void Parse::merge_common(Parse::Block* target, int pnum) {
         t = target->stack_type_at(j - tmp_jvms->stkoff());
       }
       if (t != nullptr && t != Type::BOTTOM) {
-        if (n->is_InlineType() && !t->is_inlinetypeptr()) {
-          // Allocate inline type in src block to be able to merge it with oop in target block
-          map()->set_req(j, n->as_InlineType()->buffer(this));
-        } else if (!n->is_InlineType() && t->is_inlinetypeptr()) {
-          // Scalarize null in src block to be able to merge it with inline type in target block
-          assert(gvn().type(n)->is_zero_type(), "Should have been scalarized");
-          map()->set_req(j, InlineTypeNode::make_null(gvn(), t->inline_klass()));
+        if (!t->is_inlinetypeptr()) {
+          if (n->is_InlineType()) {
+            // The merge is an oop phi, we need to buffer the inline type
+            map()->set_req(j, n->as_InlineType()->buffer(this));
+          }
+        } else {
+          Node* phi = nullptr;
+          if (target->is_merged()) {
+            phi = target->start_map()->in(j);
+          }
+          if (phi != nullptr && !phi->is_InlineType()) {
+            if (n->is_InlineType()) {
+              // The merge is an oop phi, we need to buffer the inline type
+              map()->set_req(j, n->as_InlineType()->buffer(this));
+            }
+          } else {
+            if (!n->is_InlineType()) {
+              // We cannot blindly expand an inline type here since it may be larval
+              if (gvn().type(n)->is_zero_type()) {
+                // Null constant implies that this is not a larval objects
+                map()->set_req(j, InlineTypeNode::make_null(gvn(), t->inline_klass()));
+              } else if (phi != nullptr && phi->is_InlineType()) {
+                // Larval oops cannot be merged with non-larval ones
```

Comment:
```suggestion
                // Larval oops cannot be merged with non-larval ones and since the merge point is non-larval, 'n' must be non-larval as well
```